src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs

```diff
@@ -216,14 +214,43 @@ internal static bool TryGetPathOfFileAbove(string file, DirectoryInfo startingDi
         }
 
         /// <summary>
-        /// Fires the <see cref="FileRead" /> event for the specified file.
+        /// Fires the <see cref="FileReadStart" /> event for the specified file.
         /// </summary>
         /// <param name="filePath">The full path to file that was read.</param>
-        private void OnFileRead(string filePath)
+        /// <param name="projectFullPath">The full path to project file that caused the file to be read.</param>
+        /// <param name="solutionFullPath">The full path to solution file that cause the file to be read.</param>
+        private void OnFileReadStart(string filePath, string projectFullPath, string solutionFullPath)
         {
-            EventHandler<string> fileReadEventHandler = FileRead;
+            EventHandler<GlobalJsonFileReadEventArgs> fileReadStartEventHandler = FileReadStart;
 
-            fileReadEventHandler?.Invoke(this, filePath);
+            fileReadStartEventHandler?.Invoke(
+                this,
+                new GlobalJsonFileReadEventArgs
```

Comment:
Granted we don't expect json files to be read super often, especially because of caching. However, adding an event, just so an event listener replays the event through EventSource. How about this code just calls EventSource directly instead?

Again, we don't expect this to be called very often, so this will have negligible impact, but this code is allocating a `GlobalJsonFileReadEventArgs` instance, and `OnFileReadStop` allocates a second instance, that needs to be GC'd afterwards. Even if there is nothing listening to the EventSource's events. If this class stops using events, and invokes the EventSource directly, it doesn't need to allocate these event arg instances.

Reply:
Yeah my default inclination was to keep the owner of tracing to the single component, `NuGetSdkResolver`.  But you're right, the `GlobalJsonReader` is only used in this assembly and could probably just call the tracing events directly.